src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOpAnalyzer.php

```diff
@@ -240,19 +251,117 @@ public static function analyze(
-                        'Cannot compare ' . $stmt_left_type->getId() . ' to ' . $stmt_right_type->getId(),
-                        new CodeLocation($statements_analyzer, $stmt),
-                    ),
-                    $statements_analyzer->getSuppressedIssues(),
-                );
+                if ($statements_analyzer->getCodebase()->config->strict_binary_operands
+                    && (($stmt_left_type->isSingle() && $stmt_left_type->hasBool())
+                        || ($stmt_right_type->isSingle() && $stmt_right_type->hasBool()))) {
+                    IssueBuffer::maybeAdd(
+                        new InvalidOperand(
+                            'Cannot compare ' . $stmt_left_type->getId() . ' to ' . $stmt_right_type->getId(),
+                            new CodeLocation($statements_analyzer, $stmt),
+                        ),
+                        $statements_analyzer->getSuppressedIssues(),
+                    );
+                } else {
+                    $left_literal_int_floats = self::getMinMaxLiteralIntFloat($stmt_left_type);
+                    $right_literal_int_floats = self::getMinMaxLiteralIntFloat($stmt_right_type);
+
+                    foreach ($stmt_left_type->getAtomicTypes() as $atomic_type) {
+                        if ($atomic_type instanceof TString
+                            || $atomic_type instanceof TInt
+                            || $atomic_type instanceof TArrayKey
+                            || $atomic_type instanceof TFloat
+                            || $atomic_type instanceof TResource) {
+                            continue;
+                        }
+
+                        if (AtomicTypeComparator::isContainedBy(
+                            $statements_analyzer->getCodebase(),
+                            $atomic_type,
+                            new TNamedObject('DateTimeInterface'),
+                        )) {
+                            continue;
+                        }
+
+                        if (($atomic_type instanceof TNull || $atomic_type instanceof TFalse)
+                            && !$stmt_left_type->isSingle()
+                            && $right_literal_int_floats !== []
+                            && (($stmt instanceof PhpParser\Node\Expr\BinaryOp\Greater
+                                 && min($right_literal_int_floats) >= 0)
+                                || ($stmt instanceof PhpParser\Node\Expr\BinaryOp\GreaterOrEqual
+                                    && min($right_literal_int_floats) > 0)
+                                || ($stmt instanceof PhpParser\Node\Expr\BinaryOp\Smaller
+                                    && max($right_literal_int_floats) <= 0)
+                                || ($stmt instanceof PhpParser\Node\Expr\BinaryOp\SmallerOrEqual
+                                    && max($right_literal_int_floats) < 0))
+                        ) {
+                            continue;
+                        }
```

Comment:
This is repeated (almost identically) below. We should abstract to a private function